File: scripts/4_table1_weather_analysis.R

input_dir = "D:/# Jvasco/# Portfolio/Curriculum/6. CIMMYT Scientist/_WUR PPS Wheat Trials/1-calibration-jsilva-files/Data Input/WOFOST Inputs"
output_dir = "D:/# Jvasco/# Portfolio/Curriculum/6. CIMMYT Scientist/_WUR PPS Wheat Trials/4-hberghuijs-model--final/output"

# ----------------------------------------------------------------------------------------------------------------------
# load data

stations = c('DE BILT', 'EELDE', 'VLISSINGEN')
weather_crop_veg_final <- c()
weather_crop_rep_final <- c()
weather_crop_all_final <- c()

for(s in stations){
  # weather data
  weather = readxl::read_excel(paste0(input_dir, "/KNMI Weather/# WOFOST_Weather_KNMI_", s, "_ANALYSIS.xlsx"))
  weather['DAY'] = as.Date(weather$DAY, format='%Y-%m-%d')
  weather['year'] = format(as.Date(weather$DAY, format="%Y-%m-%d"),"%Y")
  # crop data
  crop = readxl::read_excel(paste0(output_dir, "/yp-long-term-summary-", s, "-co2-clay.xlsx"))
  crop['DOE'] = as.Date(crop$DOE, format='%Y-%m-%d')
  crop['DOA'] = as.Date(crop$DOA, format='%Y-%m-%d')
  crop['DOM'] = as.Date(crop$DOM, format='%Y-%m-%d')
  for(y in unique(crop$year)){
    # subset crop per yr
    crop_yr <- subset(crop, year == y)
    doe <- crop_yr$DOE
    doa <- crop_yr$DOA
    dom <- crop_yr$DOM
    # subset weather
    weather_veg <- subset(weather, DAY > doe & DAY < doa)
    weather_rep <- subset(weather, DAY > doa & DAY < dom)
    weather_all <- subset(weather, DAY > doe & DAY < dom)
    # to data frame
    weather_crop_veg <- data.frame('station'=s, 'year'=y, 
                                   'mean_tmax'=mean(weather_veg$TMAX), 'mean_tmin'=mean(weather_veg$TMIN),
                                   'sum_rad'=sum(weather_veg$IRRAD), 'sum_rain'=sum(weather_veg$RAIN))
    weather_crop_veg_final <- rbind(weather_crop_veg_final, weather_crop_veg)
    # to data frame
    weather_crop_rep <- data.frame('station'=s, 'year'=y, 
                                   'mean_tmax'=mean(weather_rep$TMAX), 'mean_tmin'=mean(weather_rep$TMIN),
                                   'sum_rad'=sum(weather_rep$IRRAD), 'sum_rain'=sum(weather_rep$RAIN))
    weather_crop_rep_final <- rbind(weather_crop_rep_final, weather_crop_rep)
    # to data frame
    weather_crop_all <- data.frame('station'=s, 'year'=y, 
                                   'mean_tmax'=mean(weather_all$TMAX), 'mean_tmin'=mean(weather_all$TMIN),
                                   'sum_rad'=sum(weather_all$IRRAD), 'sum_rain'=sum(weather_all$RAIN))
    weather_crop_all_final <- rbind(weather_crop_all_final, weather_crop_all) } }

# ----------------------------------------------------------------------------------------------------------------------
# fit regressions

models_final_all_stations = c()
models_final_veg_stations = c()
models_final_rep_stations = c()
for(s in stations){
  # all season
  weather_crop_all_final_s <- subset(weather_crop_all_final, station == s)
  weather_crop_all_final_s$sum_rad <- weather_crop_all_final_s$sum_rad/1000 # conversion kJ to MJ
  model_tmax <- lm(mean_tmax ~ year, data=weather_crop_all_final_s)
  model_tmax <- data.frame(cbind(VariableName=rownames(summary(model_tmax)$coef), summary(model_tmax)$coef)[,c(1,2,5)])
  model_tmax$y_var <- 'mean_tmax'
  rownames(model_tmax) <- NULL
  model_tmin <- lm(mean_tmin ~ year, data=weather_crop_all_final_s)
  model_tmin <- data.frame(cbind(VariableName=rownames(summary(model_tmin)$coef), summary(model_tmin)$coef)[,c(1,2,5)])
  model_tmin$y_var <- 'mean_tmin'
  rownames(model_tmin) <- NULL
  model_srad <- lm(sum_rad ~ year, data=weather_crop_all_final_s)
  model_srad <- data.frame(cbind(VariableName=rownames(summary(model_srad)$coef), summary(model_srad)$coef)[,c(1,2,5)])
  model_srad$y_var <- 'sum_rad'
  rownames(model_srad) <- NULL
  model_rain <- lm(sum_rain ~ year, data=weather_crop_all_final_s)
  model_rain <- data.frame(cbind(VariableName=rownames(summary(model_rain)$coef), summary(model_rain)$coef)[,c(1,2,5)])
  model_rain$y_var <- 'sum_rain'
  rownames(model_rain) <- NULL
  models_final_all <- rbind(model_srad, model_tmax, model_tmin, model_rain)
  models_final_all$signif <- ifelse(models_final_all$Pr...t.. < 0.05, '*', 'ns')
  models_final_all$time <- 'all_season'
  models_final_all$station <- s
  models_final_all_stations <- rbind(models_final_all_stations, models_final_all)
  # vegetative
  weather_crop_veg_final_s <- subset(weather_crop_veg_final, station == s)
  weather_crop_veg_final_s$sum_rad <- weather_crop_veg_final_s$sum_rad/1000 # conversion kJ to MJ
  model_tmax <- lm(mean_tmax ~ year, data=weather_crop_veg_final_s)
  model_tmax <- data.frame(cbind(VariableName=rownames(summary(model_tmax)$coef), summary(model_tmax)$coef)[,c(1,2,5)])
  model_tmax$y_var <- 'mean_tmax'
  rownames(model_tmax) <- NULL
  model_tmin <- lm(mean_tmin ~ year, data=weather_crop_veg_final_s)
  model_tmin <- data.frame(cbind(VariableName=rownames(summary(model_tmin)$coef), summary(model_tmin)$coef)[,c(1,2,5)])
  model_tmin$y_var <- 'mean_tmin'
  rownames(model_tmin) <- NULL
  model_srad <- lm(sum_rad ~ year, data=weather_crop_veg_final_s)
  model_srad <- data.frame(cbind(VariableName=rownames(summary(model_srad)$coef), summary(model_srad)$coef)[,c(1,2,5)])
  model_srad$y_var <- 'sum_rad'
  rownames(model_srad) <- NULL
  model_rain <- lm(sum_rain ~ year, data=weather_crop_veg_final_s)
  model_rain <- data.frame(cbind(VariableName=rownames(summary(model_rain)$coef), summary(model_rain)$coef)[,c(1,2,5)])
  model_rain$y_var <- 'sum_rain'
  rownames(model_rain) <- NULL
  models_final_veg <- rbind(model_srad, model_tmax, model_tmin, model_rain)
  models_final_veg$signif <- ifelse(models_final_veg$Pr...t.. < 0.05, '*', 'ns')
  models_final_veg$time <- 'vegetative'
  models_final_veg$station <- s
  models_final_veg_stations <- rbind(models_final_veg_stations, models_final_veg)
  # reproductive
  weather_crop_rep_final_s <- subset(weather_crop_rep_final, station == s)
  weather_crop_rep_final_s$sum_rad <- weather_crop_rep_final_s$sum_rad/1000 # conversion kJ to MJ
  model_tmax <- lm(mean_tmax ~ year, data=weather_crop_rep_final_s)
  model_tmax <- data.frame(cbind(VariableName=rownames(summary(model_tmax)$coef), summary(model_tmax)$coef)[,c(1,2,5)])
  model_tmax$y_var <- 'mean_tmax'
  rownames(model_tmax) <- NULL
  model_tmin <- lm(mean_tmin ~ year, data=weather_crop_rep_final_s)
  model_tmin <- data.frame(cbind(VariableName=rownames(summary(model_tmin)$coef), summary(model_tmin)$coef)[,c(1,2,5)])
  model_tmin$y_var <- 'mean_tmin'
  rownames(model_tmin) <- NULL
  model_srad <- lm(sum_rad ~ year, data=weather_crop_rep_final_s)
  model_srad <- data.frame(cbind(VariableName=rownames(summary(model_srad)$coef), summary(model_srad)$coef)[,c(1,2,5)])
  model_srad$y_var <- 'sum_rad'
  rownames(model_srad) <- NULL
  model_rain <- lm(sum_rain ~ year, data=weather_crop_rep_final_s)
  model_rain <- data.frame(cbind(VariableName=rownames(summary(model_rain)$coef), summary(model_rain)$coef)[,c(1,2,5)])
  model_rain$y_var <- 'sum_rain'
  rownames(model_rain) <- NULL
  models_final_rep <- rbind(model_srad, model_tmax, model_tmin, model_rain)
  models_final_rep$signif <- ifelse(models_final_rep$Pr...t.. < 0.05, '*', 'ns')
  models_final_rep$time <- 'reproductive'
  models_final_rep$station <- s
  models_final_rep_stations <- rbind(models_final_rep_stations, models_final_rep)}
all_models <- rbind(models_final_veg_stations, models_final_rep_stations, models_final_all_stations)
  
# ----------------------------------------------------------------------------------------------------------------------
# final table

all_models <- subset(all_models, VariableName == 'year')
all_models$Estimate <- round(as.numeric(all_models$Estimate), 3)
all_models$Estimate <- paste(all_models$Estimate, all_models$signif)
all_models <- all_models[c(7,6,4,1,2)]
all_models <- reshape2::dcast(all_models, time + y_var ~ station, value.var='Estimate')
all_models <- all_models[order(all_models$y_var),]
all_models <- all_models[c(2,1,4,3,5)]
tabResultX <- xtable::xtable(all_models)
print(tabResultX, include.rownames=FALSE)

File: scripts/4_table1_weather_analysis.py
# ----------------------------------------------------------------------------------------------------------------------
# general packages

import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# ----------------------------------------------------------------------------------------------------------------------
# working directories

input_dir = r"D:\# Jvasco\# Portfolio\Curriculum\6. CIMMYT Scientist\_WUR PPS Wheat Trials\1-calibration-jsilva-files\Data Input\WOFOST Inputs"
output_dir = r"D:\# Jvasco\# Portfolio\Curriculum\6. CIMMYT Scientist\_WUR PPS Wheat Trials\4-hberghuijs-model--final\output"

# ----------------------------------------------------------------------------------------------------------------------
# load data

stations = ['DE BILT', 'EELDE', 'VLISSINGEN']
for s in stations:

    weather = pd.read_excel(os.path.join(input_dir, f".\KNMI Weather\# WOFOST_Weather_KNMI_{s}_ANALYSIS.xlsx"))
    weather['DAY'] = pd.to_datetime(weather['DAY'])
    weather['year'] = weather['DAY'].dt.year
    weather['month'] = weather['DAY'].dt.month
    wheat_months = weather[(weather.month > 1) & (weather.month < 9)]

# ------------------------------------------------------------------------------------------------------------------
    # group by month

    average = wheat_months.groupby(['month'])['TMAX', 'TMIN'].mean().round(2).reset_index()
    average_wheat = average.transpose().drop(index='month')
    average_wheat = average_wheat.rename(columns={0: 'February', 1: 'March', 2: 'April', 3: 'May', 4: 'June', 5: 'July', 6: 'August'})
    sum = wheat_months.groupby(['month', 'year'])['IRRAD', 'RAIN'].sum().round(2).reset_index()
    sum = sum.groupby(['month'])['IRRAD', 'RAIN'].mean().round(2).reset_index()
    sum['IRRAD'] = sum['IRRAD']/1000
    sum['IRRAD'] = sum['IRRAD'].round(2)
    sum_wheat = sum.transpose().drop(index='month')
    sum_wheat = sum_wheat.rename(columns={0: 'February', 1: 'March', 2: 'April', 3: 'May', 4: 'June', 5: 'July', 6: 'August'})
    wheat_final = average_wheat.append(sum_wheat)
    wheat_final.to_csv(os.path.join(output_dir, r".\weather-summary-{s}.csv".format(s=s)))

    # ------------------------------------------------------------------------------------------------------------------
    # group by year

    average = wheat_months.groupby(['year'])['TMAX', 'TMIN'].mean().round(2).reset_index()
    average_wheat = average.transpose().drop(index='year')
    sum = wheat_months.groupby(['year'])['IRRAD', 'RAIN'].sum().round(2).reset_index()
    sum = sum.groupby(['year'])['IRRAD', 'RAIN'].mean().round(2).reset_index()
    sum['IRRAD'] = sum['IRRAD']/1000
    sum['IRRAD'] = sum['IRRAD'].round(2)
    sum_wheat = sum.transpose().drop(index='year')
    wheat_final = average_wheat.append(sum_wheat)
    wheat_final.to_csv(os.path.join(output_dir, f".\weather-summary-year-{s}.csv"))

    # ------------------------------------------------------------------------------------------------------------------
    # group by year for months < 8

    left, width = .255, .71
    bottom, height = .25, .71
    right = left + width
    top = bottom + height
    wspace = 0.25
    hspace = 0.25
    kws_points = dict(s=100, alpha=0.9, linewidth=0.7)
    # ------------------------------------------------------------------------------------------------------------------
    f, (ax1, ax2) = plt.subplots(1, 2, figsize=(13.5, 6))
    axes = plt.gca()
    f.subplots_adjust(wspace=wspace, hspace=hspace)
    # ------------------------------------------------------------------------------------------------------------------
    ax1.grid(linestyle='-', linewidth=0.5, zorder=0)
    average = wheat_months.groupby(['year'])['TMAX', 'TMIN'].mean().round(2).reset_index()
    y = average.TMAX.dropna()
    X = average[["year"]].dropna()
    linr_model = LinearRegression().fit(X, y)
    linr_model.coef_
    linr_model.intercept_
    linr_model.predict(average[["year"]].dropna())
    r2 = r2_score(y_true=average.TMAX.dropna(), y_pred=linr_model.predict(average[["year"]].dropna()))
    ax1.plot(average.year, average.TMAX, color='orangered', zorder=3, label='')
    ax1.scatter(average.year, average.TMAX, color='orange', edgecolors='orangered', alpha=1, s=80, zorder=4, label='TMAX')
    ax1.plot(average.year, linr_model.predict(average[["year"]]), color="black", label='R2={r2}'.format(r2=r2.round(2)), linestyle='--', zorder=5)
    ax1.text(left + 0.675, bottom + 0.5, '{coef} degC/yr'.format(coef=linr_model.coef_[0].round(3)), color='orangered',
             horizontalalignment='right', verticalalignment='top', transform=ax1.transAxes, family='sans-serif', fontsize=17)
    y = average.TMIN.dropna()
    X = average[["year"]].dropna()
    linr_model = LinearRegression().fit(X, y)
    linr_model.coef_
    linr_model.intercept_
    linr_model.predict(average[["year"]].dropna())
    r2 = r2_score(y_true=average.TMIN.dropna(), y_pred=linr_model.predict(average[["year"]].dropna()))
    ax1.plot(average.year, average.TMIN, color='darkblue', zorder=2, label='')
    ax1.scatter(average.year, average.TMIN, color='royalblue', edgecolors='darkblue', alpha=1, s=80, zorder=3, label='TMIN')
    ax1.plot(average.year, linr_model.predict(average[["year"]]), color="black", label='R2={r2}'.format(r2=r2.round(2)), linestyle='-', zorder=5)
    ax1.text(left + 0.675, bottom + 0.025, '{coef} degC/yr'.format(coef=linr_model.coef_[0].round(3)), color='darkblue',
         horizontalalignment='right', verticalalignment='top', transform=ax1.transAxes, family='sans-serif', fontsize=17)
    ax1.set_ylim([0, 20])
    ax1.set_xlim([1970, 2020])
    ax1.set_yticks([0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20])
    ax1.set_xticklabels([1970, 1980, 1990, 2000, 2010, 2020], fontsize=15)
    ax1.set_yticklabels([0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20], fontsize=14)
    ax1.set_ylabel('Temperature (degrees C)', fontsize=15)
    ax1.set_facecolor('whitesmoke')
    ax1.legend(loc='lower right', ncol=2, fontsize=12)
    # x = average.year
    # y = average.TMAX
    # p = np.polyfit(x, y, deg=1)
    # x = average.year
    # y = p[1] + p[0] * average.year
    # ax1.plot(x, y, '-', color='black', zorder=5)
    # x = average.year
    # y = average.TMIN
    # p = np.polyfit(x, y, deg=1)
    # x = average.year
    # y = p[1] + p[0] * average.year
    # ax1.plot(x, y, '--', color='black', zorder=5)
    # ------------------------------------------------------------------------------------------------------------------
    ax2.grid(linestyle='-', linewidth=0.5, zorder=0)
    sum = wheat_months.groupby(['year'])['IRRAD', 'RAIN'].sum().round(2).reset_index()
    y = sum.IRRAD/1000
    X = sum[["year"]].dropna()
    linr_model = LinearRegression().fit(X, y)
    linr_model.coef_
    linr_model.intercept_
    linr_model.predict(average[["year"]].dropna())
    r2 = r2_score(y_true=sum.IRRAD/1000, y_pred=linr_model.predict(average[["year"]].dropna()))
    ax2.plot(sum.year, sum.IRRAD/1000, color='orangered', zorder=2, label='')
    ax2.scatter(sum.year, sum.IRRAD/1000, color='orange', edgecolors='orangered', alpha=1, s=80, zorder=3, label='Solar radiation')
    ax2.plot(sum.year, linr_model.predict(sum[["year"]]), color="black", label='R2={r2}'.format(r2=r2.round(2)), linestyle='--', zorder=5)
    ax2.text(left + 0.675, bottom + 0.515, '{coef} MJ/m2/yr'.format(coef=linr_model.coef_[0].round(1)), color='orangered',
             horizontalalignment='right', verticalalignment='top', transform=ax2.transAxes, family='sans-serif', fontsize=17)
    ax2.set_ylim([0, 3500])
    ax2.set_xlim([1970, 2020])
    ax2.set_xticklabels([1970, 1980, 1990, 2000, 2010, 2020], fontsize=15)
    ax2.set_yticklabels([0, 500, 1000, 1500, 2000, 2500, 3000, 3500], fontsize=14)
    ax2.set_ylabel('Solar radiation (MJ/m2)', fontsize=15)
    ax2.legend(loc='lower right', ncol=1, fontsize=12)

    co2 = pd.read_excel(r"D:\# Jvasco\# Portfolio\Curriculum\6. CIMMYT Scientist\_WUR PPS Wheat Trials\1-calibration-jsilva-files\Data Input\Weather Data\CO2_NOAA.xlsx")
    co2 = co2[(co2.year > 1969) & (co2.year < 2018)]
    ax3 = ax2.twinx()
    ax3.plot(co2.year, co2['mean'], color='darkblue', zorder=2, label='')
    ax3.scatter(co2.year, co2['mean'], color='royalblue', edgecolors='darkblue', alpha=1, s=80, zorder=3, label='Atmospheric CO2')
    y = co2['mean']
    X = co2[["year"]].dropna()
    linr_model = LinearRegression().fit(X, y)
    linr_model.coef_
    linr_model.intercept_
    ax3.plot(co2.year, linr_model.predict(co2[["year"]]), color="black", label='R2={r2}'.format(r2=r2.round(2)),
         linestyle='-', zorder=5)
    ax2.text(left + 0.675, bottom + 0.155, '{coef} ppm/yr'.format(coef=linr_model.coef_[0].round(1)), color='darkblue',
         horizontalalignment='right', verticalalignment='top', transform=ax2.transAxes, family='sans-serif', fontsize=17)
    ax3.set_ylim([300, 475])
    ax3.set_yticks([300, 325, 350, 375, 400, 425, 450, 475])
    ax3.set_yticklabels([300, 325, 350, 375, 400, 425, 450, 475], fontsize=14)
    ax3.set_ylabel('Atmospheric CO2 concentration (ppm)', rotation=-90, fontsize=15, labelpad=17)
    ax3.legend(loc='lower right', ncol=1, fontsize=12, bbox_to_anchor=(1, 0.125))
    ax2.set_facecolor('whitesmoke')
    # x = sum.year
    # y = sum.IRRAD/1000
    # p = np.polyfit(x, y, deg=1)
    # x = sum.year
    # y = p[1] + p[0] * sum.year
    # ax2.plot(x, y, '-', color='black', zorder=5)
    # ------------------------------------------------------------------------------------------------------------------
    f.savefig(os.path.join(output_dir, f".\weather-analysis-{s}.pdf"), bbox_inches='tight')

# ----------------------------------------------------------------------------------------------------------------------
# ----------------------------------------------------------------------------------------------------------------------
